eta/core/module.py

```diff
@@ -452,24 +453,37 @@ def is_data(self):
         '''Returns True/False if this parameter is Data.'''
         return etat.is_data(self.type)
 
+    @property
+    def default_value(self):
+        '''Gets the default value for this parameter.'''
+        if self.is_required:
+            raise ModuleMetadataError(
+                "Module parameter '%s' is required, so it has no default "
+                "value" % self.name)
+        return self._default
+
     @staticmethod
     def _parse_type(name, type_str):
         type_ = etat.parse_type(type_str)
         if not etat.is_builtin(type_) and not etat.is_concrete_data(type_):
-            raise ModuleMetadataError((
+            raise ModuleMetadataError(
                 "Module parameter '%s' has type '%s' but must be a subclass "
-                "of Builtin or ConcreteData") % (name, type_))
+                "of Builtin or ConcreteData" % (name, type_))
         return type_
 
     def _validate_default(self):
-        if self.is_builtin:
-            is_valid = self.type.is_valid_value(self.default)
+        if self._default is None:
```

Comment:
Only question regards the use of `no_default`.  If a parameter is `required and has `no_default`, then how does this bubble down to this `_validate_default` function?  It would get to the last `if not is_valid` as a `True`, but should we output a more descriptive note for this case, which is a bit different than the case when a developer or user sets it to an invalid but not the `no_default`.

Reply:
This function is only called for optional parameters, since required parameters don't have defaults